core/src/test/java/org/keycloak/sdjwt/sdjwtvp/SdJwtVPVerificationTest.java

```diff
@@ -0,0 +1,424 @@
+    }
+
+    private void testShouldFailGeneric(
+            String testFilePath,
+            KeyBindingJwtVerificationOpts keyBindingJwtVerificationOpts,
+            String exceptionMessage,
+            String exceptionCauseMessage
+    ) {
+        String sdJwtVPString = TestUtils.readFileAsString(getClass(), testFilePath);
+        SdJwtVP sdJwtVP = SdJwtVP.of(sdJwtVPString);
+
+        var exception = assertThrows(
+                VerificationException.class,
+                () -> sdJwtVP.verify(
+                        defaultIssuerSignedJwtVerificationOpts().build(),
+                        keyBindingJwtVerificationOpts
+                )
+        );
+
+        assertEquals(exceptionMessage, exception.getMessage());
+        if (exceptionCauseMessage != null) {
+            assertThat(exception.getCause().getMessage(), containsString(exceptionCauseMessage));
+        }
+    }
+
+    private void testShouldFailGenericS20(
+            JsonNode kbPayloadSubstitute,
+            KeyBindingJwtVerificationOpts keyBindingJwtVerificationOpts,
+            String exceptionMessage,
+            String exceptionCauseMessage
+    ) {
+        KeyBindingJWT keyBindingJWT = KeyBindingJWT.from(
+                kbPayloadSubstitute,
+                testSettings.holderSigContext,
+                KeyBindingJWT.TYP
+        );
+
+        String sdJwtVPString = TestUtils.readFileAsString(getClass(), "sdjwt/s20.1-sdjwt+kb.txt");
+        SdJwtVP sdJwtVP = SdJwtVP.of(
+                sdJwtVPString.substring(0, sdJwtVPString.lastIndexOf(SdJwt.DELIMITER) + 1)
+                        + keyBindingJWT.toJws()
+        );
+
+        var exception = assertThrows(
+                VerificationException.class,
+                () -> sdJwtVP.verify(
+                        defaultIssuerSignedJwtVerificationOpts().build(),
+                        keyBindingJwtVerificationOpts
+                )
+        );
+
+        assertEquals(exceptionMessage, exception.getMessage());
+        if (exceptionCauseMessage != null) {
+            assertEquals(exceptionCauseMessage, exception.getCause().getMessage());
+        }
+    }
+
+    private IssuerSignedJwtVerificationOpts.Builder defaultIssuerSignedJwtVerificationOpts() {
+        return IssuerSignedJwtVerificationOpts.builder()
+                .withVerifier(testSettings.issuerVerifierContext)
+                .withValidateIssuedAtClaim(false)
+                .withValidateNotBeforeClaim(false);
+    }
+
+    private KeyBindingJwtVerificationOpts.Builder defaultKeyBindingJwtVerificationOpts() {
+        return KeyBindingJwtVerificationOpts.builder()
+                .withKeyBindingRequired(true)
+                .withAllowedMaxAge(Integer.MAX_VALUE)
+                .withNonce("1234567890")
+                .withAud("https://verifier.example.org")
+                .withValidateExpirationClaim(false)
+                .withValidateNotBeforeClaim(false);
+    }
+
+    private ObjectNode exampleS20KbPayload() {
```

Comment:
What does `S20` means? If possible, could you name the method that other people can understand?

Reply:
I've removed the `S20` particle to make it less obscure. The method has no further intent beyond pre-constructing a key binding jwt payload.